src/Caching/StackExchangeRedis/src/RedisOutputCacheStore.cs

```diff
@@ -0,0 +1,487 @@
+    internal RedisOutputCacheStore(IOptions<RedisCacheOptions> optionsAccessor, ILogger logger)
+    {
+        ArgumentNullThrowHelper.ThrowIfNull(optionsAccessor);
+        ArgumentNullThrowHelper.ThrowIfNull(logger);
+
+        _options = optionsAccessor.Value;
+        _logger = logger;
+
+        // This allows partitioning a single backend cache for use with multiple apps/services.
+
+        // SE.Redis allows efficient append of key-prefix scenarios, but we can help it
+        // avoid some work/allocations by forcing the key-prefix to be a byte[]; SE.Redis
+        // would do this itself anyway, using UTF8
+        _valueKeyPrefix = (RedisKey)Encoding.UTF8.GetBytes(_options.InstanceName + "__MSOCV_");
+        _tagKeyPrefix = (RedisKey)Encoding.UTF8.GetBytes(_options.InstanceName + "__MSOCT_");
+        _tagMasterKey = (RedisKey)Encoding.UTF8.GetBytes(_options.InstanceName + "__MSOCT");
+        _tagMasterKeyArray = new[] { _tagMasterKey };
+
+        _ = Task.Factory.StartNew(RunGarbageCollectionLoopAsync, default, TaskCreationOptions.LongRunning, TaskScheduler.Current);
+    }
+
+    private async Task RunGarbageCollectionLoopAsync()
+    {
+        try
+        {
+            while (!Volatile.Read(ref _disposed))
+            {
+                // approx every 5 minutes, with some randomization to prevent spikes of pile-on
+                var secondsWithJitter = 300 + Random.Shared.Next(-30, 30);
+                Debug.Assert(secondsWithJitter >= 270 && secondsWithJitter <= 330);
+                await Task.Delay(TimeSpan.FromSeconds(secondsWithJitter)).ConfigureAwait(false);
+                try
+                {
+                    if (GarbageCollectionEnabled)
+                    {
+                        await ExecuteGarbageCollectionAsync(GetExpirationTimestamp(TimeSpan.Zero)).ConfigureAwait(false);
+                    }
+                }
+                catch (Exception ex)
+                {
+                    // this sweep failed; log it
+                    _logger.LogDebug(ex, "Transient error occurred executing redis output-cache GC loop");
+                }
+            }
+        }
+        catch (Exception ex)
+        {
+            // the entire loop is dead
+            _logger.LogDebug(ex, "Fatal error occurred executing redis output-cache GC loop");
+        }
+    }
+
+    internal async ValueTask<long> ExecuteGarbageCollectionAsync(long keepValuesGreaterThan, CancellationToken cancellationToken = default)
+    {
+        var cache = await ConnectAsync(CancellationToken.None).ConfigureAwait(false);
+
+        var gcKey = _tagMasterKey.Append("GC");
+        var gcLifetime = TimeSpan.FromMinutes(5);
+        // value is purely placeholder; it is the existence that matters
+        if (!await cache.StringSetAsync(gcKey, DateTime.UtcNow.ToString(CultureInfo.InvariantCulture), gcLifetime, when: When.NotExists).ConfigureAwait(false))
+        {
+            return -1; // signal competition
```

Comment:
"completion"?

Reply:
no, competition from another app instance, and we were unable to set the key, because it already existed

Reply:
Consider using a `Tuple<bool, long?>` here rather than a magic value?

Reply:
In fact, it might be worth collecting some metrics here around how many cache entries there are in total and how many were collected. It would probably mean that you'd need to either use another script to get the count of entries and then remove the entries - or two calls.

I'm just thinking if there is any other way we could get a complete picture of actual cache population across all instances of the application.

Reply:
Slight problem there, @mitchdenny : we can count *unique tags* trivially, and we can count *value tags* as part of the GC sweep (although that will double/triple/etc count when values have multiple tags), but it is impossible in redis to count the most important metrics "number of entries" and "space of entries". Well, not impossible, but absurdly expensive - you'd need to do a full `SCAN` sweep just to count, and getting space requires a metadata query per entry. Neither of these is recommended on production systems.

Reply:
@mitchdenny re the value-tuple: we cross-target netfx, iirc. Value-tuple on netfx is part of the "assembly binding trifecta of fail" - friends don't let friends use value-tuples on netfx

Reply:
magic numbers: resolved; switched from `long` to `long?` for the result, with `null` being the "we didn't even try"